### PR TITLE
Sym knn

### PR DESCRIPTION
**Pull request by michalk8**
Hey @Marius1311 ,
I've added the functionality, keeping the original intact. I've rebased this branch to use current scvelo's master (I've kept this master the same), that's why you see Volker's commits.
In `scvelo/preprocessing/moments.py` you can see that I differientiate which method to call when asking whether to recompute neighbors, based on whether use_rep is None or not.
It's because in original neighbors_to_be_recomputed, the else branch also contains the following check n_neighs.max() * .1 > n_neighs.min() . I wanted to be most backwards compatible, so I've opted out for this option.
Let me know what you think.

### CODE REVIEW

**Comment by Marius1311** [comment]
Thanks Mike. This looks good, I will have a more thourough look later.

File: scvelo/preprocessing/neighbors.py [hunk 1]
```diff
@@ -207,6 +208,23 @@ def neighbors_to_be_recomputed(adata, n_neighbors=None):
...
+                     or 'params' not in adata.uns['neighbors'] \
+                     or (n_neighbors is not None and n_neighbors != adata.uns['neighbors']['params']['n_neighbors']) \
+                     or (use_rep is not None and use_rep != adata.uns['neighbors']['params'].get('use_rep', None))
+    if invalid_neighs:
+        return True
+
+    neighbor_params = adata.uns['neighbors']['params']
+    logg.info('using the precomputed KNN graph with {} neighbors computed in basis {!r}'.
+              format(neighbor_params['n_neighbors'], neighbor_params.get('use_rep', None)), r=True)
+
```
Comment (Marius1311) [inline scvelo/preprocessing/neighbors.py:224]: There is a conceptual difference here in that this function wants n_neighbors to match exactly and it recomputes if they don't. I actually like this behaviour and think this is the right thing to do.

**Comment by Marius1311** [comment]
All works fine, there is one interesting thing I realised. Say I run the following commends:

```python
sc.pp.neighbors(adata, n_neighbors=40)
scv.pp.moments(adata)
scv.tl.velocity_graph_sym(adata, n_neighbors=30)
```

In that case, the `velocity_graph_combined` and the `connectivities` will differ in my particular data example (ldg) by two elements. That's because of the following slightly complicated reason:
- When the moments are computed, since no basis is given, it runs `neighbors_to_be_recomputed`. This only checks whether the given number of neighbors is smaller than what has been computed. 
- When the indices are obtained, these are pruned such that each cell has *exactly* 30 neighbors. In other words, the original K=40 KNN graph is pruned back to one where each cell has precisely 30 neighbors
- As it happens, two cells which are identical in their 30 neirest neighbors now have precisely the same neighbors
- This means that these two cells will be absolutely identical in the `.layers['Ms']` due to the way moments are computed (cells which have the same neighbors will get the same value)
- When we now compute the velocity graph, we compute `cos (x_i-x_j, v_i)`. Note that x_i and x_j are the profiles of cell i and j in `.layers['Ms']. 
- However, this is defined to be zero because x_i and x_j are identical. 
- For that reason, the velocity graph will be zero here, while the connectivities are nonzero.

This can also happen if the moments are computed in an embedding, in such a case, 2 cells are much more likely to have exactly the same neighbors. 

The above isn't too dramatic and I think we can leave things as they are, I just wanted to spell it out in case we wonder later on what's going on here. If one removes the cells which are identical in `.layers['Ms']`, then the two graphs will have exactly the same non-zero elements again.

**Comment by Marius1311** [comment]
Do you think we are ready to open a PR to main scvelo repo? We will have to explain well why this is needed.

**Comment by michalk8** [comment]
> Do you think we are ready to open a PR to main scvelo repo? We will have to explain well why this is needed.

I think we can, if you don't want to modify it any further (e.g. by adding option to remove or just always remove identical cells).

**Comment by Marius1311** [comment]
I think it's fine as it is now, lets rather kick off the discussion via a PR to scvelo. Can you rebase this branch so it has all of Volker's recent commits to upstream master and start a PR to upstream scvelo master? Please mention me in the PR, I will then add some more explanation to this.

**Comment by michalk8** [comment]
> I think it's fine as it is now, lets rather kick off the discussion via a PR to scvelo. Can you rebase this branch so it has all of Volker's recent commits to upstream master and start a PR to upstream scvelo master? Please mention me in the PR, I will then add some more explanation to this.

Made a PR just now.

**Comment by Marius1311** [comment]
Oh, I remember this.... good that we are done with that.

**Pull request closed without merging**